gobblin-modules/gobblin-orc/src/main/java/org/apache/gobblin/writer/GobblinBaseOrcWriter.java

```diff
@@ -153,14 +157,18 @@ public GobblinBaseOrcWriter(FsDataWriterBuilder<S, D> builder, State properties)
 
     // Create value-writer which is essentially a record-by-record-converter with buffering in batch.
     this.inputSchema = builder.getSchema();
-    TypeDescription typeDescription = getOrcSchema();
+    this.typeDescription = getOrcSchema();
     this.valueWriter = getOrcValueWriter(typeDescription, this.inputSchema, properties);
     this.batchSize = properties.getPropAsInt(ORC_WRITER_BATCH_SIZE, DEFAULT_ORC_WRITER_BATCH_SIZE);
-    this.rowBatch = typeDescription.createRowBatch(this.batchSize);
+    this.rowBatchPool = RowBatchPool.instance(properties);
```

Comment:
I did some digging, and it seems like we use the GobblinOrcWriter in 2 spots. 
1. [Fork](https://github.com/apache/gobblin/blob/022b49fdd99cebcc21aad9b6b7cde9df0b092537/gobblin-runtime/src/main/java/org/apache/gobblin/runtime/fork/Fork.java#L249)
2. [AbstractJobLauncher](https://github.com/apache/gobblin/blob/b726a606cea3deb567b1fdeeba9acbcc220e6d30/gobblin-runtime/src/main/java/org/apache/gobblin/runtime/AbstractJobLauncher.java#L516)

In (1) we init a writer per fork and we may have multiple forks in a Task, especially in streaming model task runner. From my understanding, a fork is long running so we only create a new orc writer during the the initial run of a task and retries
In (2) we init once during the job launch. 

I think I am missing something here though because we wouldn't need to recycle the row batches via a pool if they were create once and forget. 🤔

Reply:
Yes, there are not created once. A new OrcBaseWriter is created whenever we flush a file in FI which is happening every 5 mins. Also  a new ORC writer will be created if we are writing 2+ partitions at once. See `PartitionedDataWriter`

Reply:
Considering the frequency of recycles, what's your opinion on https://www.baeldung.com/java-singleton-double-checked-locking for the singleton. Reduces the need for as many lock accesses